README.md
[![Build Status](https://travis-ci.org/cartazio/streaming-machine-json.svg?branch=master)](https://travis-ci.org/cartazio/streaming-machine-json)

# Properly streaming incremental json parser

CHANGELOG.md
# Revision history for streaming-machine-json

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/MyLibTest.hs
{-# LANGUAGE OverloadedStrings,ScopedTypeVariables #-}
module Main (main) where


import InternalJSON
import Data.Machine.Type (runT)
import qualified Data.ByteString.Char8 as BSC
import TokenJSON
import Control.Exception (SomeException)



referenceLiterals :: [[BSC.ByteString]]
referenceLiterals = referenceNotObjects ++ referenceObjects


referenceObjects :: [[BSC.ByteString]]
referenceObjects= [
                    ["[","]"]
                    ,["{}"]
                   ,["{\"a\": 0.0 }"]
                   ,["{","}"]
                   ,["[1,\"c\",[]]"]
                ]

referenceNotObjects   :: [[BSC.ByteString]]
referenceNotObjects = [ ["[]"]
                    ,["[0.0]"]
                   ,[ "0.0"]
                   ,["0.","0"]
                   ,["0",".0"]
                   ,["[{\"a\" : 21}, \"hello\\u29E71\\t\nworld\"]"]
                   ,["true","false","null"]
                   ]


refNotObjectResults :: [[JsonToken]]
refNotObjectResults = [[EmptyArray],[OpenDelim Array,TokenNumber 0.0,CloseDelim Array]
                    ,[TokenNumber 0.0],[TokenNumber 0.0],[TokenNumber 0.0]
                    ,[OpenDelim Array,OpenDelim Object,TokenText "a",TokenSeparator Colon,TokenNumber 21.0
                              ,CloseDelim Object,TokenSeparator Comma,TokenText "hello\10727\&1\t\nworld",CloseDelim Array]
                    ,[TokenBool True,TokenBool False,TokenNull]]



refObjectResults :: [[JsonToken]]
refObjectResults = [[EmptyArray]
                    ,[EmptyObject]
                    ,[OpenDelim Object,TokenText "a",TokenSeparator Colon,TokenNumber 0.0,CloseDelim Object]
                    ,[EmptyObject]
                    ,[OpenDelim Array,TokenNumber 1.0,TokenSeparator Comma,TokenText "c",TokenSeparator Comma,EmptyArray,CloseDelim Array] ]

refObjectArrayDigest :: [[(SnocList Accessor, PrimVal)]]
refObjectArrayDigest  =[[(RNil,PrimEmptyArray)]
                        ,[(RNil,PrimEmptyObject)]
                        ,[((:|) RNil (ObjectField "a"),PrimNumber 0.0)],[(RNil,PrimEmptyObject)]
                        ,[((:|) RNil (ArrayIx 0),PrimNumber 1.0),((:|) RNil (ArrayIx 1),PrimString "c"),((:|) RNil (ArrayIx 2),PrimEmptyArray)]]

main :: IO ()
main = do

  outputTokens <- mapM (\ str -> runT ({-batchHarnessMachina-} batchChunkedParse str))
                   referenceLiterals
  print outputTokens

  if outputTokens /= ( refNotObjectResults ++
                      refObjectResults
                    )
    then  error $ "bad program\n "  ++ show outputTokens ++ "\n"
    else return ()
  putStrLn "and now we look at the digestion"
  outputStreams:: Either SomeException [[(SnocList Accessor, PrimVal)]] <- return $  mapM  (\ str -> runT (batchHarnessMachina str))
                      referenceObjects
  putStrLn "now we look at those summaries"
  print outputStreams
  case outputStreams of
      Right os ->    if os  /=   refObjectArrayDigest
                       then error "bad outputStreams digest!"
                       else return ()
  return ()


src/InternalJSON.hs
{-# LANGUAGE DeriveDataTypeable  #-}
{-# LANGUAGE DeriveFoldable      #-}
{-# LANGUAGE DeriveFunctor       #-}
{-# LANGUAGE DeriveGeneric       #-}
{-# LANGUAGE DeriveTraversable   #-}
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE RankNTypes          #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE FlexibleContexts #-}

module InternalJSON
    (
     Accessor (..)
    , PrimVal (..)
    , SnocList (..)
    , eventor
    , parserState
    , batchFeed
    , batchHarness
    , batchChunkedParse
    , batchHarnessMachina
    , chunkedParse
    , snoc2List
    , list2Snoc
    ,JError(..)
    ) where

import Control.Applicative
import Control.Monad.Trans.Class
-- import Control.Monad.Except
import Control.Monad.Catch
import Control.Exception(Exception,SomeException)

import Data.Data (Data)
import Data.Typeable (Typeable)

import GHC.Generics (Generic)

import qualified Data.Attoparsec.ByteString.Char8 as ABC
import qualified Data.ByteString.Char8 as BSC

import Data.Machine


import Control.Monad.ST.Strict (ST)
import InternalParserJSON
import  StreamJSON

-- import Data.Resampler.Types.Error


data JError =
    BadState    [JsonToken]  (SnocList DelimSort)  (SnocList Accessor )
  | LexingError String
  | EarlyTerm   [JsonToken]  (SnocList DelimSort)  (SnocList Accessor )
  | InitOrFinal  JsonToken
  | InvalidSeq  [JsonToken]  (SnocList DelimSort)  (SnocList Accessor)
  deriving(Eq,Ord,Show,Data,Typeable)

instance Exception JError

snoc2List :: SnocList a -> [a]
snoc2List =  reverse . raw2List
  where
    raw2List (as :| a) = a : raw2List as
    raw2List RNil = []


list2Snoc :: [a] -> SnocList a
list2Snoc = snocReverse . list2raw
  where
    list2raw (a : as) = list2raw  as :| a
    list2raw [] = RNil


snocReverse :: SnocList a -> SnocList a
snocReverse = snRev  RNil
  where
    snRev ls RNil = ls
    snRev ls (as :| a) = snRev (ls :| a) as


-- Convert a parser for `a` into a model that accepts `Bytestring` chunks and
-- produces values of type `a`.
{-# SPECIALIZE  chunkedParseGeneral ::  ABC.Parser a -> ProcessT IO BSC.ByteString a#-}
{-# SPECIALIZE  chunkedParseGeneral ::  ABC.Parser a -> ProcessT (Either SomeException) BSC.ByteString a#-}
{-# SPECIALIZE  chunkedParseGeneral ::  ABC.Parser a -> ProcessT (ST s) BSC.ByteString a#-}
chunkedParseGeneral :: forall m a  . MonadThrow  m => ABC.Parser a -> ProcessT m BSC.ByteString a
chunkedParseGeneral parser = construct $ await >>=   startParse parser


startParse :: forall m a b . MonadThrow m =>  ABC.Parser a -> BSC.ByteString -> PlanT (Is BSC.ByteString) a m b
startParse parser = \bs -> start bs

  where
    start :: BSC.ByteString -> PlanT (Is BSC.ByteString) a m b
    start bs = do
      result <- ABC.parseWith (await <|> return BSC.empty) parser bs
      case result of
        ABC.Fail remains contexts message ->  lift $ throwM $ LexingError $ "ChunkedParseGeneral: "++message++" remains: " ++ show remains ++ "contexts: " ++ show contexts
        ABC.Partial _ -> error "chunkedParseGeneral: parseWith produced Partial."
        ABC.Done remains res -> yield res >>
                                if BSC.null remains then  stop else start remains

-- Process `Bytestring` chunks into `JsonToken`s.
{-# SPECIALIZE  chunkedParse :: ProcessT IO BSC.ByteString JsonToken#-}
chunkedParse :: forall   m . MonadThrow  m => ProcessT m BSC.ByteString JsonToken
chunkedParse = chunkedParseGeneral $
               do
                 _<-ABC.many' ABC.space
                 p  <-  parseTokenPrimValue
                 _<- ABC.many' ABC.space
                 return p


token2JPrimVal:: JsonToken ->Maybe PrimVal
token2JPrimVal (TokenText t)   = Just $ PrimString t
token2JPrimVal (TokenBool b)   = Just $ PrimBool b
token2JPrimVal (TokenNumber n) = Just $ PrimNumber n
token2JPrimVal (TokenNull)     = Just PrimNull
token2JPrimVal EmptyArray      = Just PrimEmptyArray
token2JPrimVal EmptyObject     = Just PrimEmptyObject
token2JPrimVal (OpenDelim _ )  = Nothing
token2JPrimVal (CloseDelim _ )  = Nothing
token2JPrimVal (TokenSeparator _ )  = Nothing

-- TODO: should empty arrays and objects be primvalues?
-- TODO: We think that ... "f":,"g": ... will sneak through
{-# SPECIALIZE eventor :: ProcessT IO JsonToken (SnocList Accessor,PrimVal) #-}
{-# SPECIALIZE eventor :: ProcessT (Either SomeException) JsonToken (SnocList Accessor,PrimVal) #-}
{-# SPECIALIZE eventor :: ProcessT (ST s) JsonToken (SnocList Accessor,PrimVal) #-}
-- | 'eventor' acts as a token stream validator, for valid JSON objects. If you
-- cant assume
eventor :: forall m e . MonadThrow   m
        => ProcessT m JsonToken (SnocList Accessor,PrimVal)
eventor = construct $ parserState [] RNil RNil


parserState :: forall m res .
              MonadThrow m =>
              [JsonToken]
            -> SnocList DelimSort
            -> SnocList Accessor
            -> PlanT (Is JsonToken) (SnocList Accessor,PrimVal) m res

parserState (TokenSeparator Comma : rest) e2@(_ :| Open Array) (acslist :| ArrayIx i) =
    parserState rest e2 (acslist :| ArrayIx (i+1))

parserState (OpenDelim Object : TokenText t : TokenSeparator Colon : rest) dlist acslist =
    parserState rest (dlist :| Open Object) (acslist :| ObjectField t)

parserState (OpenDelim Array: rest) dlist acslist =
    parserState rest (dlist :| Open Array) (acslist :| ArrayIx 0)

-- the close delim leads are for nested cases
parserState e1@(CloseDelim Object : rest) e2@(dlist :| dhead) e3@(acslist :| achead) =
    case (dhead, achead) of
      (Open Object, ObjectField _) -> parserState rest dlist acslist
      _                            -> lift $ throwM $ BadState e1 e2 e3

parserState e1@(CloseDelim Array : rest) e2@(dlist :| dhead) e3@(acslist :| achead) =
    case (dhead,achead) of
      (Open Array, ArrayIx _) -> parserState rest dlist acslist
      _                       -> lift $ throwM $ BadState e1 e2 e3

parserState e1@(primv : TokenSeparator Comma : rest) e2@(_ :| Open Array) e3@(acslist :| ArrayIx i)
  | (Just v) <- token2JPrimVal primv = yield (e3,v) >> parserState rest e2 (acslist :| ArrayIx (i+1))
  | otherwise                        = lift $ throwM $ BadState e1 e2 e3

parserState e1@(primv : CloseDelim Array : rest)
            e2@(dlist:| Open Array)
            e3@(acslist :| ArrayIx _i)
  | (Just v) <- token2JPrimVal primv = yield (e3,v) >> parserState rest dlist acslist
  | otherwise                        = lift $ throwM $ BadState e1 e2 e3

parserState e1@(primv : CloseDelim Object : rest)
            e2@(dlist :| Open Object)
            e3@(acslist :| ObjectField _t)
  | (Just v) <- token2JPrimVal primv = yield (e3,v) >> parserState rest dlist acslist
  | otherwise                        = lift $ throwM $ BadState e1 e2 e3
parserState [] RNil RNil = await >>= initOrFinalDelim

parserState e1@(primv : TokenSeparator Comma : TokenText t : TokenSeparator Colon : rest)
            e2@(_dlist :| Open Object)
            e3@(acslist :| ObjectField _f)
  | (Just v) <- token2JPrimVal primv = yield (e3,v) >> parserState rest e2 (acslist :| ObjectField t)
  | otherwise                        = lift $ throwM  $ BadState e1 e2 e3

parserState _e1@(TokenSeparator Comma : TokenText t : TokenSeparator Colon : rest)
            e2@(_dlist :| Open Object)
            _e3@(acslist :| ObjectField _f)
  = parserState rest e2 (acslist :| ObjectField t)

parserState context nesting path
  | length context <= 4 = (await <|> (lift . throwM $ EarlyTerm context nesting path)) >>= (\tok -> parserState (context ++ [tok]) nesting path)
  | otherwise = lift $ throwM  $ InvalidSeq context nesting path

initOrFinalDelim :: MonadThrow m => JsonToken -> PlanT (Is JsonToken) (SnocList Accessor, PrimVal) m res
initOrFinalDelim od
 | od == OpenDelim Object || od == OpenDelim Array = parserState [od] RNil RNil
  ---- NOTE, i'm very suspicious of these two, but it seems morally the right choice
 | od == EmptyObject  = yield (RNil,PrimEmptyObject) >> stop
 | od == EmptyArray  = yield (RNil,PrimEmptyArray) >> stop
initOrFinalDelim od = lift $ throwM $ InitOrFinal od




-- Testing Functions --
batchFeed :: [a] -> Source a
batchFeed x = construct $ mapM_ yield x >> stop

batchHarness :: MonadThrow  m
             => [BSC.ByteString] -> m [(SnocList Accessor,PrimVal)]
batchHarness x = runT $ batchHarnessMachina x

batchChunkedParse :: MonadThrow  m => [BSC.ByteString] -> MachineT m k JsonToken
batchChunkedParse x = batchFeed x ~> chunkedParse


batchHarnessMachina :: MonadThrow  m =>
     [BSC.ByteString]
     -> MachineT m k (SnocList Accessor, PrimVal)
batchHarnessMachina x = batchFeed x ~> chunkedParse ~>  eventor

src/InternalParseJSON.hs
{-# OPTIONS_GHC -funbox-strict-fields #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE CPP #-}
{-# LANGUAGE OverloadedStrings #-}


module InternalParseJSON(
  JsonToken(..)
  ,Separator(..)
  ,ConType(..)
  ,parseTokenPrimValue
  ,DelimSort(..)
  ) where

import Data.ByteString.Builder
  (Builder, byteString, toLazyByteString, charUtf8, word8)
import Control.Applicative (empty,(<|>))
import Data.Attoparsec.ByteString.Char8 (char,Parser,string,peekChar',double)
import Data.Bits ((.|.), shiftL)
import Data.ByteString (ByteString)
import Data.Char (chr, isDigit)
import Data.Text (Text)
import Data.Text.Encoding (decodeUtf8')

import qualified Data.Attoparsec.ByteString as A
import qualified Data.Attoparsec.Zepto as Z
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Unsafe as B

import TokenJSON


#define BACKSLASH 92
#define CLOSE_CURLY 125
#define CLOSE_SQUARE 93
#define COMMA 44
#define DOUBLE_QUOTE 34
#define OPEN_CURLY 123
#define OPEN_SQUARE 91
#define C_0 48
#define C_9 57
#define C_A 65
#define C_F 70
#define C_a 97
#define C_f 102
#define C_n 110
#define C_t 116


-- | 'parseTokenValue' will parse any "token" value, namely true,false,String,Number, or Null
parseTokenPrimValue :: Parser JsonToken
parseTokenPrimValue =  do
    nextChar <- peekChar'
    case nextChar of
      '[' -> ( string "[]" >> return EmptyArray )  <|> (do _<-  char '[' ; return $! OpenDelim Array)
      ']' -> do  _<- char ']'; return $! CloseDelim Array
      '{' -> (do _<-  string "{}" ; return EmptyObject) <|> (do  _ <-char '{' ; return $! OpenDelim Object)
      '}' -> do _<- char '}' ; return $! CloseDelim Object
      ':' -> do _<-  char ':' ; return $! TokenSeparator Colon
      ',' -> do _ <- char ',' ; return $! TokenSeparator Comma
      '"' -> fmap (TokenText $!) jstring  -- Do we really need $! here?
      't' -> do _ <- string "true" ; return $! TokenBool  True
      'f' -> do _ <- string "false" ; return $! TokenBool False
      'n' -> do _<- string "null" ; return TokenNull
      c  | c == '+' || c == '-' || isDigit c ->   fmap TokenNumber double
                                                    --- REPLACE DOUBLE WITH SCIENTIFIC Parse (and eager decimal for 0.0 corner case)
      _ -> empty


-- | Parse a top-level JSON value.  This must be either an object or
-- an array, per RFC 4627.
--
-- The conversion of a parsed value to a Haskell value is deferred
-- until the Haskell value is needed.  This may improve performance if
-- only a subset of the results of conversions are needed, but at a
-- cost in thunk allocation.


-- | Parse a quoted JSON string.
jstring :: Parser Text
jstring = A.word8 DOUBLE_QUOTE *> jstring_

-- | Parse a string without a leading quote.
jstring_ :: Parser Text
jstring_ = {-# SCC "jstring_" #-} do
  s <- A.scan False $ \s c -> if s then Just False
                                   else if c == DOUBLE_QUOTE
                                        then Nothing
                                        else Just (c == BACKSLASH)
  _ <- A.word8 DOUBLE_QUOTE
  s1 <- if BACKSLASH `B.elem` s
        then case Z.parse unescape s of
            Right r  -> return r
            Left err -> fail err
         else return s

  case decodeUtf8' s1 of
      Right r  -> return r
      Left err -> fail $ show err

{-# INLINE jstring_ #-}

unescape :: Z.Parser ByteString
unescape = toByteString <$> go mempty where
  go acc = do
    h <- Z.takeWhile (/=BACKSLASH)
    let rest = do
          start <- Z.take 2
          let !slash = B.unsafeHead start
              !t = B.unsafeIndex start 1
              escape = case B.findIndex (==t) "\"\\/ntbrfu" of
                         Just i -> i
                         _      -> 255
          if slash /= BACKSLASH || escape == 255
            then fail "invalid JSON escape sequence"
            else do
            let cont m = go (acc `mappend` byteString h `mappend` m)
                {-# INLINE cont #-}
            if t /= 117 -- 'u'
              then cont (word8 (B.unsafeIndex mapping escape))
              else do
                   a <- hexQuad
                   if a < 0xd800 || a > 0xdfff
                     then cont (charUtf8 (chr a))
                     else do
                       b <- Z.string "\\u" *> hexQuad
                       if a <= 0xdbff && b >= 0xdc00 && b <= 0xdfff
                         then let !c = ((a - 0xd800) `shiftL` 10) +
                                       (b - 0xdc00) + 0x10000
                              in cont (charUtf8 (chr c))
                         else fail "invalid UTF-16 surrogates"
    done <- Z.atEnd
    if done
      then return (acc `mappend` byteString h)
      else rest
  mapping = "\"\\/\n\t\b\r\f"

hexQuad :: Z.Parser Int
hexQuad = do
  s <- Z.take 4
  let hex n | w >= C_0 && w <= C_9 = w - C_0
            | w >= C_a && w <= C_f = w - 87
            | w >= C_A && w <= C_F = w - 55
            | otherwise          = 255
        where w = fromIntegral $ B.unsafeIndex s n
      a = hex 0; b = hex 1; c = hex 2; d = hex 3
  if (a .|. b .|. c .|. d) /= 255
    then return $! d .|. (c `shiftL` 4) .|. (b `shiftL` 8) .|. (a `shiftL` 12)
    else fail "invalid hex escape"


toByteString :: Builder -> ByteString
toByteString = L.toStrict . toLazyByteString
{-# INLINE toByteString #-}

src/InternalParserJSON.hs
{-# OPTIONS_GHC -funbox-strict-fields #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE CPP #-}
{-# LANGUAGE OverloadedStrings #-}

module InternalParserJSON(
  JsonToken(..)
  ,Separator(..)
  ,ConType(..)
  ,parseTokenPrimValue
  ,DelimSort(..)
  ) where

import Data.ByteString.Builder
  (Builder, byteString, toLazyByteString, charUtf8, word8)
import Control.Applicative (empty,(<|>))
import Data.Attoparsec.ByteString.Char8 (char,Parser,string,peekChar',double)
import Data.Bits ((.|.), shiftL)
import Data.ByteString (ByteString)
import Data.Char (chr, isDigit)
import Data.Text (Text)
import Data.Text.Encoding (decodeUtf8')

import qualified Data.Attoparsec.ByteString as A
import qualified Data.Attoparsec.Zepto as Z
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import qualified Data.ByteString.Unsafe as B


-- these imports are only needed pre GHC 7.10 / pre
#if !MIN_VERSION_base(4,8,0)
import Data.Monoid
import Control.Applicative
#endif
----


import TokenJSON


#define BACKSLASH 92
#define CLOSE_CURLY 125
#define CLOSE_SQUARE 93
#define COMMA 44
#define DOUBLE_QUOTE 34
#define OPEN_CURLY 123
#define OPEN_SQUARE 91
#define C_0 48
#define C_9 57
#define C_A 65
#define C_F 70
#define C_a 97
#define C_f 102
#define C_n 110
#define C_t 116


-- | 'parseTokenValue' will parse any "token" value, namely true,false,String,Number, or Null
parseTokenPrimValue :: Parser JsonToken
parseTokenPrimValue =  do
    nextChar <- peekChar'
    case nextChar of
      '[' -> ( string "[]" >> return EmptyArray )  <|> (do _<-  char '[' ; return $! OpenDelim Array)
      ']' -> do  _<- char ']'; return $! CloseDelim Array
      '{' -> (do _<-  string "{}" ; return EmptyObject) <|> (do  _ <-char '{' ; return $! OpenDelim Object)
      '}' -> do _<- char '}' ; return $! CloseDelim Object
      ':' -> do _<-  char ':' ; return $! TokenSeparator Colon
      ',' -> do _ <- char ',' ; return $! TokenSeparator Comma
      '"' -> fmap (TokenText $!) jstring  -- Do we really need $! here?
      't' -> do _ <- string "true" ; return $! TokenBool  True
      'f' -> do _ <- string "false" ; return $! TokenBool False
      'n' -> do _<- string "null" ; return TokenNull
      c  | c == '+' || c == '-' || isDigit c ->   fmap TokenNumber double
                                                    --- REPLACE DOUBLE WITH SCIENTIFIC Parse (and eager decimal for 0.0 corner case)
      _ -> empty


-- | Parse a top-level JSON value.  This must be either an object or
-- an array, per RFC 4627.
--
-- The conversion of a parsed value to a Haskell value is deferred
-- until the Haskell value is needed.  This may improve performance if
-- only a subset of the results of conversions are needed, but at a
-- cost in thunk allocation.


-- | Parse a quoted JSON string.
jstring :: Parser Text
jstring = A.word8 DOUBLE_QUOTE *> jstring_

-- | Parse a string without a leading quote.
jstring_ :: Parser Text
jstring_ = {-# SCC "jstring_" #-} do
  s <- A.scan False $ \s c -> if s then Just False
                                   else if c == DOUBLE_QUOTE
                                        then Nothing
                                        else Just (c == BACKSLASH)
  _ <- A.word8 DOUBLE_QUOTE
  s1 <- if BACKSLASH `B.elem` s
        then case Z.parse unescape s of
            Right r  -> return r
            Left err -> fail err
         else return s

  case decodeUtf8' s1 of
      Right r  -> return r
      Left err -> fail $ show err

{-# INLINE jstring_ #-}

unescape :: Z.Parser ByteString
unescape = toByteString <$> go mempty where
  go acc = do
    h <- Z.takeWhile (/=BACKSLASH)
    let rest = do
          start <- Z.take 2
          let !slash = B.unsafeHead start
              !t = B.unsafeIndex start 1
              escape = case B.findIndex (==t) "\"\\/ntbrfu" of
                         Just i -> i
                         _      -> 255
          if slash /= BACKSLASH || escape == 255
            then fail "invalid JSON escape sequence"
            else do
            let cont m = go (acc `mappend` byteString h `mappend` m)
                {-# INLINE cont #-}
            if t /= 117 -- 'u'
              then cont (word8 (B.unsafeIndex mapping escape))
              else do
                   a <- hexQuad
                   if a < 0xd800 || a > 0xdfff
                     then cont (charUtf8 (chr a))
                     else do
                       b <- Z.string "\\u" *> hexQuad
                       if a <= 0xdbff && b >= 0xdc00 && b <= 0xdfff
                         then let !c = ((a - 0xd800) `shiftL` 10) +
                                       (b - 0xdc00) + 0x10000
                              in cont (charUtf8 (chr c))
                         else fail "invalid UTF-16 surrogates"
    done <- Z.atEnd
    if done
      then return (acc `mappend` byteString h)
      else rest
  mapping = "\"\\/\n\t\b\r\f"

hexQuad :: Z.Parser Int
hexQuad = do
  s <- Z.take 4
  let hex n | w >= C_0 && w <= C_9 = w - C_0
            | w >= C_a && w <= C_f = w - 87
            | w >= C_A && w <= C_F = w - 55
            | otherwise          = 255
        where w = fromIntegral $ B.unsafeIndex s n
      a = hex 0; b = hex 1; c = hex 2; d = hex 3
  if (a .|. b .|. c .|. d) /= 255
    then return $! d .|. (c `shiftL` 4) .|. (b `shiftL` 8) .|. (a `shiftL` 12)
    else fail "invalid hex escape"


toByteString :: Builder -> ByteString
toByteString = L.toStrict . toLazyByteString
{-# INLINE toByteString #-}